sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -93,6 +106,15 @@ case class CoalesceShufflePartitions(session: SparkSession) extends CustomShuffl
     }
   }
 
+  private def shouldApplyChildren(plan: SparkPlan): Boolean = {
+    plan.find(p => shouldApplyChildrenFunc(p)).isDefined
+  }
+
+  private def shouldApplyChildrenFunc(plan: SparkPlan): Boolean = plan match {
+    case _: UnionExec => true
```

Comment:
Any other plan node that we can apply the same optimization into? If no, could you inline it?
```
  private def shouldApplyChildren(plan: SparkPlan): Boolean = {
    plan.find(_.isInstanceOf[Union]).isDefined
  }
```
Then, `shouldApplyChildren` -> `hasUnion`?